### DISCUSSION TITLE
nonzero() like nonnegative() and nonpositive()

### DISCUSSION THREAD

**Discussion by topperspal** [colinhacks/zod#1776] [Q&A] [answered]
I want to parse mongodb sort object like this
```js
{
  name: 1,
  age: 1
}
```
how can I ensure that the value is one of [-1, 1], I know, I can use it with different approaches like using `preprocess` or `transform` but they are not one-word solutions (**zod is preferable for**). If nonzero() is implemented, we can write this like this
```ts
const validator = z.object({
  age: z.number().min(-1).max(1).nonzero(),
});
```

This can also achieved with `in([...])` modifier
```ts
const validator = z.object({
  age: z.number().in([-1,1]),
});
```

The `in([...])` modifier can be found in many validators. `nonzero()` is also very useful()

**Comment by igalklebanov** [comment]
>```ts
>const validator = z.object({
>  age: z.number().min(-1).max(1).nonzero(),
>});
>```

I believe this allows non-integers in range (e.g. `0.5`). Pretty sure that's not what you had in mind.

A better, real-world, example would be a money transaction, the amount is either negative or positive.

```ts
const money = z.object({
  amount: z.number().nonzero(),
  currency: z.string().length(3),
})
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Are these what you are looking for?

```ts
const nonZeroSchema = z.number().positive().or( z.number().negative() )
console.log( nonZeroSchema.safeParse( 42 ).success ) // true
console.log( nonZeroSchema.safeParse( -42 ).success ) // true
console.log( nonZeroSchema.safeParse( 0.5 ).success ) // true
console.log( nonZeroSchema.safeParse( -0.5 ).success ) // true
console.log( nonZeroSchema.safeParse( 0 ).success ) // false

const oneOrMinusOneSchema = z.literal( 1 ).or( z.literal( -1 ) )
console.log( oneOrMinusOneSchema.safeParse( 1 ).success ) // true
console.log( oneOrMinusOneSchema.safeParse( -1 ).success ) // true
console.log( oneOrMinusOneSchema.safeParse( 0.5 ).success ) // false
console.log( oneOrMinusOneSchema.safeParse( -0.5 ).success ) // false
console.log( oneOrMinusOneSchema.safeParse( 0 ).success ) // false
```

Maybe I'm misunderstanding, but these seem fairly easy to make without a dedicated method.

**Comment by topperspal** [comment]
```js
// This does not fit in this case, because we want only -1 (negative) or 1 (positive)
const nonZeroSchema = z.number().positive().or( z.number().negative() )
console.log( nonZeroSchema.safeParse( 42 ).success ) // true
```

YES! 
```js
const oneOrMinusOneSchema = z.literal( 1 ).or( z.literal( -1 ) )
```
This will do the job, but lets compare simplicity of some methods for achieving the same result

**Your approach -**
```js 
const oneOrMinusOneSchema = z.literal( 1 ).or( z.literal( -1 ) )
console.log( oneOrMinusOneSchema.safeParse( num ).success ) // true
```
but **if block** seems simple
```js
if( num === -1 || num === 1) { ... }
```
A little simpler
```js
const nonZeroSchema = z.number().min(-1).max(1).nonzero();
```
**Simplest**
```js
const nonZeroSchema = z.number().in([-1,1]);
```
This also allow us to increase accepted values in future, like if we just want to accept one of **[ 400, 401,403,404 ]**, we can use **in()** method for this
```js
const nonZeroSchema = z.number().in([ 400, 401,403,404 ]);
```


Now lets discuss **nonzero()** - 
**nonzero()** is actually required as an alternative to `z.number().positive().or( z.number().negative() )` 
```js
z.number().positive().or( z.number().negative() ); // allow either positive or negative
// or 
z.number().nonzero(); // allow either positive or negative
z.number().int().nonzero();
```

**Reply by jonblack** [reply]
`z.number().positive().or(z.number().negative() );` gives the error message `Number must be greater than zero`, which is misleading. To correct this, the solution becomes even wilder: `z.number().positive({message: "Number must be non-zero"}).or( z.number().negative() );`

I'm also in favour of a dedicated `nonzero()` method. Even without the error message, `nonzero()` shows the intent much more clearly.

**Comment by vladfaust** [comment]
I think a dedicated `nonzero()` method would be a good addition.
